### PR TITLE
HPCC-21886 Support OUTPUT, EXTEND, NAMED within a LIBRARY

### PR DESCRIPTION
**Pull request by richardkchapman**
Note: Requires that an output with the same name exists within the calling
query, and that the type matches.

Uses special sequence numbers for outputs within libraries.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
New test added

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review/merge

File: roxie/ccd/ccdserver.cpp [hunk 1]
```diff
@@ -21308,34 +21308,50 @@ class CRoxieServerWorkUnitWriteActivity : public CRoxieServerInternalSinkActivit
...
+        {
+            IConstWorkUnit *workunit = serverContext->queryQueryFactory()->queryWorkUnit();
+            assertex(workunit);
+            const char *storedName = helper.queryName();
+            assertex(storedName);
+            Owned<IConstWUResult> queryRes =  workunit->getQueryResultByName(storedName);
+            if (!queryRes)
+                throw makeStringExceptionV(0, "Library cannot write to result %s that does not exist in calling query", storedName);
+            Owned<IConstWUResult> libraryRes =  factory->queryQueryFactory().queryWorkUnit()->getResultByNameAndSequence(storedName, helper.getSequence());
+            if (libraryRes)
```
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp]: What does it mean if libraryRes is null?  Should it ever occur legally?

File: roxie/ccd/ccdserver.cpp [hunk 2]
```diff
@@ -21308,34 +21308,50 @@ class CRoxieServerWorkUnitWriteActivity : public CRoxieServerInternalSinkActivit
...
+        if (sequence >= LibraryBaseSequence)
+        {
+            IConstWorkUnit *workunit = serverContext->queryQueryFactory()->queryWorkUnit();
+            assertex(workunit);
+            const char *storedName = helper.queryName();
+            assertex(storedName);
+            Owned<IConstWUResult> queryRes =  workunit->getQueryResultByName(storedName);
+            if (!queryRes)
+                throw makeStringExceptionV(0, "Library cannot write to result %s that does not exist in calling query", storedName);
+            Owned<IConstWUResult> libraryRes =  factory->queryQueryFactory().queryWorkUnit()->getResultByNameAndSequence(storedName, helper.getSequence());
```
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp]: I think this could use getResultBySequence() with the new scheme.

File: testing/regress/ecl/aaalibrary6.ecl [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+//nohthor
+//nothor
+//nothorlcr
+//publish
+
+#option ('targetService', 'aaaLibrary5');
```
Comment (ghalliday) [inline testing/regress/ecl/aaalibrary6.ecl]: Should have been aaaLibrary6 I think

**Review by ghalliday — COMMENTED** [review]
@richardkchapman a couple of comments

File: roxie/ccd/ccdserver.cpp (hunk 1 above)
Comment (richardkchapman) [reply roxie/ccd/ccdserver.cpp]: I don't think it should occur

File: roxie/ccd/ccdserver.cpp (hunk 2 above)
Comment (richardkchapman) [reply roxie/ccd/ccdserver.cpp]: Yes, it probably could

File: testing/regress/ecl/aaalibrary6.ecl (hunk 3 above)
Comment (richardkchapman) [reply testing/regress/ecl/aaalibrary6.ecl]: I don't know what that line means - I think it was your code?

**Comment by richardkchapman** [comment]
Approved verbally by @ghalliday

**Pull request merged**